crates/swc_ecma_transforms_module/src/system_js.rs

```diff
@@ -0,0 +1,1224 @@
+    import_ident_names: Vec<JsWord>,
+}
+
+pub fn system_js(root_mark: Mark, config: Config) -> impl Fold {
+    SystemJs {
+        root_mark,
+        resolver: None::<(NoopImportResolver, _)>,
+        config,
+
+        declare_var_idents: vec![],
+        export_map: HashMap::new(),
+        export_names: vec![],
+        export_values: vec![],
+        export_identifier: quote_ident!("_export"),
+        context_identifier: quote_ident!("_context"),
+        tla: false,
+        enter_async_fn: false,
+        enter_fn: false,
+        root_fn_decl_idents: vec![],
+        module_item_meta_list: vec![],
+        import_ident_names: vec![],
+    }
+}
+
+pub fn system_js_with_resolver<R>(
+    resolver: R,
+    base: FileName,
+    root_mark: Mark,
+    config: Config,
+) -> impl Fold
+where
+    R: ImportResolver,
+{
+    SystemJs {
+        root_mark,
+        resolver: Some((resolver, base)),
+        config,
+
+        declare_var_idents: vec![],
+        export_map: HashMap::new(),
+        export_names: vec![],
+        export_values: vec![],
+        export_identifier: quote_ident!("_export"),
+        context_identifier: quote_ident!("_context"),
+        tla: false,
+        enter_async_fn: false,
+        enter_fn: false,
+        root_fn_decl_idents: vec![],
+        module_item_meta_list: vec![],
+        import_ident_names: vec![],
+    }
+}
+
+struct ModuleItemMeta {
+    export_names: Vec<JsWord>,
+    export_values: Vec<Expr>,
+    has_export_all: bool,
+    src: JsWord,
+    stmts: Vec<Stmt>,
+}
+
+impl<R> SystemJs<R>
+where
+    R: ImportResolver,
+{
+    fn export_call(&self, name: JsWord, span: Span, expr: ExprOrSpread) -> CallExpr {
+        CallExpr {
+            span,
+            callee: Callee::Expr(Box::new(Expr::Ident(self.export_identifier.clone()))),
+            args: vec![Expr::Lit(Lit::Str(quote_str!(name))).as_arg(), expr],
+            type_args: None,
+        }
+    }
+
+    fn fold_module_name_ident(&mut self, ident: Ident) -> Expr {
+        if ident.sym.eq("__moduleName") && ident.span.ctxt().outer().eq(&self.root_mark) {
+            return Expr::Member(MemberExpr {
+                span: DUMMY_SP,
+                obj: Box::new(Expr::Ident(self.context_identifier.clone())),
+                prop: MemberProp::Ident(quote_ident!("id")),
+            });
+        }
+        Expr::Ident(ident)
+    }
+
+    fn replace_assign_expr(&mut self, assign_expr: AssignExpr) -> Expr {
+        match &assign_expr.left {
+            PatOrExpr::Expr(pat_or_expr) => match &**pat_or_expr {
+                Expr::Ident(ident) => {
+                    for (k, v) in self.export_map.iter() {
+                        if is_eq_sym((ident.sym.clone(), Some(ident.span.ctxt())), k) {
+                            let mut expr = Expr::Assign(assign_expr);
+                            for value in v.iter() {
+                                expr = Expr::Call(self.export_call(
+                                    value.clone(),
```

Comment:
This is super inefficient.

Reply:
I'm not sure what resolve it, Can you help me ? I want to transform `var d = 3; export { d as e, d as f };` to `_export("f", _export("e", d = 3));`.

Reply:
Oh, sorry. It's fine, as the type is `JsWord`.